### ISSUE TITLE
Bug in `@parallel`?

### ISSUE THREAD

**Issue by nickeubank** [JuliaLang/julia#23450]
Replicating example:

Launch julia as: `julia -p 3`

Run:

     # Make sure workers present
     @everywhere(println(myid())

     # Run in parallel
     a = @parallel (+) for x in 1:6
           return 1
     end

     # check result
    println("result of adding 1 six times is: $a")

My output:

              # Make sure workers present
              @everywhere(println(myid()))
          	From worker 2:	2
         	From worker 4:	4
        	From worker 3:	3

              # Run in parallel
              a = @parallel (+) for x in 1:6
                    return 1
              end

              # check result
              println("result of adding 1 six times is: $a")
              result of adding 1 six times is: 3

Expected: 

              result of adding 1 six times is: 6

Version Info:

    julia> versioninfo()
    Julia Version 0.6.0
    Commit 903644385b (2017-06-19 13:05 UTC)
    Platform Info:
      OS: macOS (x86_64-apple-darwin13.4.0)
      CPU: Intel(R) Core(TM) i7-3615QM CPU @ 2.30GHz
      WORD_SIZE: 64
      BLAS: libopenblas (USE64BITINT DYNAMIC_ARCH NO_AFFINITY Sandybridge)
      LAPACK: libopenblas64_
      LIBM: libopenlibm
      LLVM: libLLVM-3.9.1 (ORCJIT, ivybridge)

EDIT: for context, not problem with all parallelism. With `pmap`:

     julia> pmap(x -> x, 1:6)
    6-element Array{Int64,1}:
    1
    2
    3
    4
    5
    6

**Comment by sbromberger** [comment]
You're using `return` which will exit all blocks. Try this:

```
a = @parallel (+) for x in 1:6
              1
        end
```

**Comment by nickeubank** [comment]
ahh, I thought since we're feeding output into the specified aggregator (`+`) a return value was required (or at least allowed). Seems odd to use implicitly returned values not now allow them to be explicit...

(@sbromberger I feel like you've become my personal Julia tutor!)

**Comment by sbromberger** [comment]
> I feel like you've become my personal Julia tutor!

Then you're in some serious trouble! :)
